### DISCUSSION TITLE
Ability to switch current playback/queue from speaker to speaker

### DISCUSSION THREAD

**Discussion by fernplant** [music-assistant/support#702] [Feature requests and ideas]
This is a feature I miss from using Amazon Music or other Google cast-able apps. I can't switch from one speaker (or group) to another while music is playing and have it pick back up wherever it was already at

Love this project, use it everyday now. Thank you

**Comment by DutchJaFO** [comment]
Spotify itself allows for something similar when ungrouping speakers I get the option to resume playback.
It would be nice to have the option to move the active output to a different speaker.
This would further improve the use case for people with multiple speaker systems.

It probably could be used to create a 'music walk with me' automation where play moves to mobile phone output when leaving the house and then back to the speaker system upon return. At the very least it may make writing such a script easier ...

**Comment by rml1111** [comment]
Would be great to have the ability to transfer music/queue playing on one speaker/speaker group in Music Assistant to another speaker/speaker group. This would be similar to [stream transfer](https://www.blog.google/products/google-nest/move-your-music-stream-transfer/) functionality in the Google Home app for Chromecast speakers. 

Absolutely love Music Assistant for Home Assistant. For me, it was the missing "killer app" for Home Assistant.

**Reply by OzGav** [reply]
I believe this is already possible. If you cast to a group then turn down the volume of all the speakers where you aren’t aren’t. When you move  turn up that speaker and turn down the other.

**Reply by rml1111** [reply]
This isn't exactly the functionality I was describing, but is a really interesting work-around in the interim. The functionality that I was describing is essentially what the Google Home app provides (the ability to select or deselect speakers for the currently playing media source. Described further in this HA forum: [Google Cast Media Transfer From One Speaker To Another in Home Assistant](https://community.home-assistant.io/t/google-cast-media-transfer-from-one-speaker-to-another-in-home-assistant/297297). Will give your suggestion a try, as I think similar functionality could be simulated via multiple automations triggered by Lovelace buttons. Thank you.

**Reply by rml1111** [reply]
I explored this over the weekend and realized that there are a couple of significant limitations to the "whole house" cast group and automations to mute or turn up volume approach. Specifically, it would take over my Chromecast enabled AV receiver any time I play music. If someone else is watching the television (connected to the receiver) it would switch the input from video source to Chromecast, resulting in the inability to watch a movie in one room and listen to music in another. Another limitation, which could potentially be worked around with even more automations is the ability to play from two different music sources on different Chromecast speakers in different rooms.

**Reply by marcelveldt** [reply]
Here's how I solved it:

- have a cast group per floor (or section you play at the same time)
- for the section that includes the AV receiver just have another group with the rest of the speakers, excluding the AV receiver.
- create a HA group with all cast groups (yes you can have groups within groups)
- maintain the queue on that "whole house" HA group player
- Use your automations to turn "on" the right cast group player.

A few tricks:

- You can just use the turn on/off controls on the underlying "child" players, MA will automatically translate that into a mute command.
- If you turn off the last child of a group, MA will auto turn off the cast group
- If you turn off the last cast group player, MA will automatically turn off the "whole house" group player.


So, to sum up;

- I have an automation per speaker to turn it on/off and set the correct volume based on time of the day, motion etc.
- I have a script to turn on the (correct) cast group player.
- I have an automation to fill the queue of the "whole house" player during the day with playlists suitable for the time of day.

In the end a lot of workarounds but it works pretty great actually. The best would be if we can group/ungroup cast speakers from MA and service calls but Google did not yet expose an api for that.

**Reply by rml1111** [reply]
Thank you for the detailed workaround instructions. I will give it a try. If I'm following correctly, this is the key functionality that will prevent my AV receiver from switching over (assuming it is part of a group that is turned off): "If you turn off the last child of a group, MA will auto turn off the cast group". Would you be able to share the scrip to turn on cast group players?

**Comment by bbigras** [comment]
Is this a duplicate of #553 ?

**Reply by OzGav** [reply]
Yes pretty much...

**Comment by tortho** [comment]
Why would we like to stream to all media players and then mute them??
- Wasted bandwidth
- speakers falling out if their not in sync
- speakers might drop out of the group if they become unavailable for a little time and will then not be readded without another automation going to capture this

Today I have the automations controlling the lights for each room. Same automation checks if the "follow me" helper is on and if so,  run a script which adjust the volume equal to the media player(s) playing and then adding the speaker to the group.
Same goes for turning off.

In the same way I do announcements, they will then hit the audience in their respective rooms.

The script adding/removing always keep 1 speaker as master (the last one or a default if something has happened) the list of players are then stored in an input select helper.

Have an issued using this in music assitant though.. The helper have the master speaker as state and list of all players including the master in the helpers options.
I cant see that I can get the helper displayed or selected as a player.

With other solutions / cards / automation I can use the following to address the speaker(s) "active":
media_player.{{states('input_select.sonos_music_master_speaker') }}

Anyone with any good thoughts of "adding" this "media player" and also to get a button on the music assistant screen with a "follow me" button. (Can section the display thogh and add a rown with buttons on the side of the iframe card..

**Comment by OzGav** [comment]
This has been implemented in 2.3.0beta 4

**Reply by tbrasser** [reply]
So it also transfers the currently playing item? Nice!

Also, is there a way to call this from HA?

**Reply by OzGav** [reply]
Soon

**Comment by rml1111** [comment]
This is awesome and thank you for implementing! I've been waiting for this since my previous post back in 2022. Once I transfer to another queue, is there a way to have it automatically start playing?

**Reply by OzGav** [reply]
Go to the now playing view and click in the menu top right

**Comment by OzGav** [comment]
The feature request has been implemented successfully. As such this feature request will be closed and locked. If there are enhancements desired then open a new [feature request](https://github.com/orgs/music-assistant/discussions/categories/feature-requests-and-ideas). If there is a question about how this works then check the [documentation](https://beta.music-assistant.io/), or ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord. If there are problems then ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord or open an [issue](https://github.com/music-assistant/support/issues). Thanks for your interest in seeing this idea being brought to life.
